sos/collector/sosnode.py

```diff
@@ -645,6 +645,10 @@ def finalize_sos_cmd(self):
                     "--namespaces=%s" % self.opts.namespaces
                 )
 
+        if self.check_sos_version('4.5.1'):
```

Comment:
That's correct. The reason this is set to 4.5.1 is that 4.5.0 has already been released, and we don't consider running from checkout when dealing with these checks for what the on-node sos supports. 4.5.1 is the earliest we can absolutely guarantee that a node will have `--low-priority` support.

Reply:
Oh, sorry, I misread your query.

Since this is running in `sosnode`, this check is being run on each node we're connected to, not (just) the node running `sos collect`.

Reply:
I see. I thought that was the intention because it's the thing which makes sense. But looking through the code I don't think this check is running on the remote, but locally. I went ahead and tested it (see below), but the rationale is that the `PackageManager` on the `Policy` class isn't being forwarded the `remote_exec` parameter. I'm talking about [this line](https://github.com/sosreport/sos/blob/11763c36ced8dd07d7586d59255e8d74754efe4d/sos/policies/__init__.py#L136).

To test this I patched the check_sos_version method [here](https://github.com/sosreport/sos/blob/11763c36ced8dd07d7586d59255e8d74754efe4d/sos/collector/sosnode.py#L422) adding:

```python
with open('/tmp/version', 'w') as f:
   f.write(_ver)
```

The file never showed up. Patching the same way on the host where I'm running `sos collect` from indeed writes the file.

I can file a different bug with this if what I'm saying above looks correct to you.

Reply:
> I see. I thought that was the intention because it's the thing which makes sense. But looking through the code I don't think this check is running on the remote, but locally. I went ahead and tested it (see below), but the rationale is that the `PackageManager` on the `Policy` class isn't being forwarded the `remote_exec` parameter. I'm talking about [this line](https://github.com/sosreport/sos/blob/11763c36ced8dd07d7586d59255e8d74754efe4d/sos/policies/__init__.py#L136).
> 

That's because it's loaded in the override [like here for Red Hat distros](https://github.com/sosreport/sos/blob/main/sos/policies/distros/redhat.py#L60).



> To test this I patched the check_sos_version method [here](https://github.com/sosreport/sos/blob/11763c36ced8dd07d7586d59255e8d74754efe4d/sos/collector/sosnode.py#L422) adding:
> 
> ```python
> with open('/tmp/version', 'w') as f:
>    f.write(_ver)
> ```
> 
> The file never showed up. Patching the same way on the host where I'm running `sos collect` from indeed writes the file.
> 
> I can file a different bug with this if what I'm saying above looks correct to you.


Version-dependent option handling is working as expected:

```
[root@[redacted] ~]# rpm -q sos
sos-4.1-9.el8_5.noarch


# On my local workstation 

[jake@terra sos]$ ./bin/sos collect --no-local --cluster-type=none --nodes=10.1.183.177 --namespaces=thisoptionshouldbeskipped -vv --password --cmd-timeout=10000 -o kernel

sos-collector (version 4.5.0)

[...]

[[redacted]:_load_sos_info] sos version is 4.1.0-9
[[redacted]:run_command] Running command sos report -l
[[redacted]:run_command] Running command sos report --list-presets

Beginning collection of sosreports from 1 nodes, collecting a maximum of 4 concurrently

[[redacted]:finalize_sos_cmd] Final sos command set to /usr/sbin/sos report --batch  --chroot auto --only-plugins=kernel
[redacted] : Generating sos report...
[[redacted]:run_command] Running command /usr/sbin/sos report --batch  --chroot auto --only-plugins=kernel
```

With the above `collect` command, we would expect a final command of `sos report -o kernel --cmd-timeout=10000 --namespaces=thisoptionshouldbeskipped`. However, sos-4.1 does not support the latter two options - [so we don't pass them on using the sosnode.check_sos_version functionality](https://github.com/sosreport/sos/blob/11763c36ced8dd07d7586d59255e8d74754efe4d/sos/collector/sosnode.py#L632-#L646)

Reply:
> The file never showed up. Patching the same way on the host where I'm running sos collect from indeed writes the file.

AH, I understand the discrepancy now.  "The check is being run on the node" in the sense of "We are checking the node's installed sos version that is set on that node's `SoSNode` object", not "we are executing this specific code remotely".

We determine the version of the remote sos installation via a remotely executed command. Then the _comparison_ is done locally between version strings so we can determine what the final sos command (which will be executed remotely) should be.